src/language-server/project/rover/DocumentSynchronization.ts

```diff
@@ -229,4 +251,48 @@ export class DocumentSynchronization {
       position: match.position,
     });
   }
+
+  handlePartDiagnostics(params: PublishDiagnosticsParams) {
+    const uriDetails = splitUri(params.uri);
+    if (!uriDetails) {
+      return;
+    }
+    const found = this.knownFiles.get(uriDetails.uri);
+    if (!found) {
+      return;
+    }
+    const part = found.parts.find(
+      (p) => p.fractionalIndex === uriDetails.fractionalIndex,
+    );
+    if (!part) {
+      return;
+    }
+    part.diagnostics = params.diagnostics;
+
+    const fullDocumentParams: PublishDiagnosticsParams = {
+      uri: found.full.uri,
+      version: found.full.version,
+      diagnostics: found.parts.flatMap((p) =>
+        p.diagnostics.map((diagnostic) => ({
+          ...diagnostic,
+          range: rangeInContainingDocument(p.source, diagnostic.range),
+        })),
+      ),
+    };
+
+    this.sendDiagnostics(fullDocumentParams);
+  }
+
+  get openDocuments() {
+    return [...this.knownFiles.values()].map((f) => f.full);
+  }
+
+  clearAllDiagnostics() {
```

Comment:
If the project is discarded (or recreated), we need to tell VSCode to reset all diagnostics reported by it - after all, the `vscode-graphql` LSP is just a facade in front of many Client or Rover projects, and VSCode doesn't know what belongs to which of those projects.